fs: enable custom mount points for fatfs

kica-z opened the pull request:
In order to use custom named mount points with fatfs some adjustments to the zephyr fatfs integration had to be done.

If somebody sees a possibility how `CONFIG_FS_FATFS_CUSTOM_MOUNT_POINT_COUNT` could be removed i am very open to suggestions.

de-nordic commented on modules/fatfs/zephyr_fatfs_config.h:
```diff
@@ -117,6 +117,20 @@
+#error Mount point count needs to be defined
+#endif /* !defined(CONFIG_FS_FATFS_CUSTOM_MOUNT_POINT_COUNT) */
+#undef FF_VOLUMES
+#define FF_VOLUMES CONFIG_FS_FATFS_CUSTOM_MOUNT_POINT_COUNT
```
You can probably use the ARG_COUNT macro for that.

kica-z replied:
Could you elaborate what you mean by that? I have not found a macro with such a name.

de-nordic replied:
Sorry, NUM_VA_ARGS. ARG_COUNT was my macro, it does not exist in Zephyr.

kica-z replied:
@de-nordic Ah ok. But it is a string so i guess that's not possible, correct? At least i have not yet found a possibility to "explode" the string at compile time.

de-nordic replied:
The FF_VOLUME_STRS is supposed to be comma separated list of strings, something like:
![image](https://github.com/user-attachments/assets/7a1fa36a-2f2c-4bba-9c6f-e9ea40454596)
https://github.com/zephyrproject-rtos/fatfs/blob/427159bf95ea49b7680facffaa29ad506b42709b/include/ffconf.h#L174

kica-z replied:
Yes but my suggestion would be to make it definable in kconfig in order to be able to have custom mount point names. The kconfig CONFIG_FS_FATFS_CUSTOM_MOUNT_POINTS replaces FF_VOLUME_STRS

de-nordic replied:
Ah, right. You can not set Kconfig to  "Hello", "Bu" , it will be "Hello, bu".
You are right, we have to have both Kconfigs. It is a pity that ELM FAT does not allow to define Volume matching function.

kica-z replied:
I reverted the change. It is now dependent on the kconfig again.

kica-z committed:
fs: enable custom mount points for fatfs

Introduce two new kconfig options in order
to be able to define custom named fatfs
mount points. If activated replace the static
FF_VOLUME_STRS approach with the runtime
generated VolumeStr array containing those
mount points.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -277,6 +277,28 @@ config FS_FATFS_MULTI_PARTITION
+
+config FS_FATFS_MOUNT_POINTS
+	string "Support for custom mountpoints when using fatfs"
+	default "RAM,NAND,CF,SD,SD2,USB,USB2,USB3"
```
I know this may be stupid, but have you checked whether something like this:
"RAM\0NAND\0CF\0SD\0"
would work?
This indeed looks awful, but you can then have this string in rdata section, and only assign to VolumeStr[...] pointers to strings within this definition.

kica-z replied:
@de-nordic I tried this approach. Unfortunately the `\0`s get converted to ASCII String 0 so they cannot be used as delimiters.

de-nordic replied:
Ok. Thanks for at least checking.
The idea was to get what your loop does but at compile time.

kica-z replied:
I don't think that would have worked even with the `\0`s as the `VolumeStr` array needs to be filled with the pointers 🤔

de-nordic replied:
Yes. But, you have string like that:
"AAA\0BBB\0CCC"
right? You know the length of string by the size of allocated data, in above case it is 3 * 3 + 3, the null terminator after one is implicit.
So now all you have to do is to assign to VolumeStr pointer array pointers to beginning of this string and sub-strings that begin after each 0 separator. You do not have to modify string as it already terminator.

Drawback of this: the Kconfig string looks bad.

kica-z replied:
Correct but that assigning of pointers still needs a loop.

But the string can be `const` as you said and does not need to be adjusted. I think we mean the same thing but were just wording it differently.

de-nordic replied:
Yes, we can not weasel out of a loop.

de-nordic replied:
The only thing that would change for us is they the string needs to be modified in runtime, so it lands in .data section and is copied from to ram; if we would make it with 0, then we can directly point into .rdata, so we save some ram.

kica-z commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -277,6 +277,28 @@ config FS_FATFS_MULTI_PARTITION
+	default 8
+	range 1 10
+	help
+	  This option has to be used in combination with FS_FATFS__MOUNT_POINTS.
```
Adjusted help string

kica-z commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -277,6 +277,28 @@ config FS_FATFS_MULTI_PARTITION
 
+config FS_FATFS_MOUNT_POINT_COUNT
+	int "Count of custom mount points"
+	default 8
```
I renamed and adjusted the kconfig options and made all code dependent on them with `#if`

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -114,8 +112,7 @@ static uint8_t translate_flags(fs_mode_t flags)
 
-static int fatfs_open(struct fs_file_t *zfp, const char *file_name,
-		      fs_mode_t mode)
+static int fatfs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t mode)
```
Can you move unrelated changes to at least separate commit?
I know that there are a few badly formatted lines to fix, but please do not do these fixes, with the same commit, until they touch your logic.

There is simple reason for that: somebody does changes over that unrelated line, then we figure that the previous commit was bad and need to revert it for a time, and due to unrelated change we have to revert two commits or entire stack of them.

kica-z replied:
I reverted them. I know that formatting should be done in a separate commit but my auto-formatter doesn't 😅.

de-nordic replied:
> I reverted them. I know that formatting should be done in a separate commit but my auto-formatter doesn't 😅.

You could still have them as separate commit. Anyway, maybe next time.
Btw. I do not know how many tricks you know with git, so sorry if I am trying to be smart here, but you can use
```
git reset --mixed <rev before commit you want to split>
git add --patch // to select changes you want in a commit
git commit -C <rev of commit you have split>
```
and you will end up with selected changes in your commit with the same message as previously; then you can add them and create new commit if you want.
I do that a lot.

kica-z replied:
😮 i did not know that. that could come in handy, thanks 👍
Probablyi will do the formatting in https://github.com/zephyrproject-rtos/zephyr/pull/84147 as i have to rebase it anyway after this one is merged.

de-nordic replied:
One more hint, before doing any changes on your branch, or experimenting, you may want to use `git branch -c` to make a copy of it, easier than tagging and dealing with reflog to fix stuff if things go wrong.

de-nordic commented:
Btw, aside from unrelated changes the rest is ok.

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -557,9 +557,30 @@ static const struct fs_file_system_t fatfs_fs = {
+	int mount_point_count = 0;
+
+	VolumeStr[0] = mount_points;
+	for (int i = 0; i < ARRAY_SIZE(mount_points) - 1; i++) {
```
> You'll probably want to handle the case (assert?) where the user has provided a count greater than the actual number of mount points described in the comma separated list?

That would be a punishment for wasting space, while providing not enough just logs error.
Warning here would be better, but yes we could communicate issue here.
